src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,317 @@
+        private static readonly SyntaxGenerator s_generator = CSharpSyntaxGenerator.Instance;
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var identifier = await root.SyntaxTree.GetTouchingTokenAsync(context.Span.Start,
+                token => token.Parent.IsKind(SyntaxKind.LocalFunctionStatement), cancellationToken).ConfigureAwait(false);
+            if (identifier == default)
+            {
+                return;
+            }
+
+            if (context.Span.Length > 0 &&
+                context.Span != identifier.Span)
+            {
+                return;
+            }
+
+            var localFunction = (LocalFunctionStatementSyntax)identifier.Parent;
+            if (localFunction.Identifier != identifier)
+            {
+                return;
+            }
+
+            if (!localFunction.Parent.IsKind(SyntaxKind.Block, out BlockSyntax parentBlock))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(c => UpdateDocumentAsync(root, document, parentBlock, localFunction, c)));
+        }
+
+        private static async Task<Document> UpdateDocumentAsync(
+            SyntaxNode root,
+            Document document,
+            BlockSyntax parentBlock,
+            LocalFunctionStatementSyntax localFunction,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var declaredSymbol = (IMethodSymbol)semanticModel.GetDeclaredSymbol(localFunction, cancellationToken);
+
+            var dataFlow = semanticModel.AnalyzeDataFlow(
+                localFunction.Body ?? (SyntaxNode)localFunction.ExpressionBody.Expression);
+            var captures = dataFlow.Captured;
+
+            // First, create a parameter per each capture so that we can pass them as arguments to the final method
+            // Filter out `this` because it doesn't need a parameter, we will just make a non-static method for that
+            // We also make a `ref` parameter here for each capture that is being written into inside the funciton
+            var capturesAsParameters = captures
+                .Where(capture => !capture.IsThisParameter())
+                .Select(capture => CodeGenerationSymbolFactory.CreateParameterSymbol(
+                    attributes: default,
+                    refKind: dataFlow.WrittenInside.Contains(capture) ? RefKind.Ref : RefKind.None,
```

Comment:
❗️ This may need to be `out`:

```csharp
    public void M() {
        bool a;
        
        Initialize();
        Console.WriteLine(a);
        
        void Initialize() => a = default;
    }
```

❓ Does invocation of a value type count as writing?

```csharp
public static void M() {
  MutableInt a = default(MutableInt);
  
  SetValue(3);
  Console.WriteLine(a.Value); // should print 3
  
  void SetValue(int value) => a.Value = value;
}

struct MutableInt
{
  public int Value;
}
```

Reply:
Test added for mutating struct

Reply:
@alrz Can you add this one as well?

```csharp
public static int M()
{
    MutableInt a;

    SetValue(3);
    return a.Value; // should print 3

    void SetValue(int value) => a.Value = value;
}

struct MutableInt
{
    public int Value;
}
```